### PR TITLE
Update requirements 

### PR DESCRIPTION
**Pull request by jdeschamps**
I tested all notebooks with TF 2.7 and 2.10 (latest), and CSBDeep 0.7.2 (latest) in Python 3.9.

- Lifted dependency requirement for CSBDeep in order to use recent TF versions
- Fixed tests and made files creation happen in temporary folders
- Solved deprecated warnings (tifffile.imsave, np.int)
- Removed travis yaml
- Added tox (tox.yml, .toml) files to run automated testing
- Added github action workflow (to be tested), including deployment to PyPI

Here is roughly how I installed the environment:
```bash
conda install -c conda-forge cudatoolkit=11.2 cudnn
export LD_LIBRARY_PATH=$LD_LIBRARY_PATH:$CONDA_PREFIX/lib/
pip install tensorflow
pip install -e .
```

### CODE REVIEW

File: setup.py [hunk 1]
```diff
@@ -33,26 +32,27 @@
...
-          "matplotlib",
-          "six",
-          "keras>=2.1.1,<2.4.0",
-          "tifffile>=2020.5.11",
+          "tifffile",
           "imagecodecs>=2020.2.18",
-          "tqdm",
           "backports.tempfile;python_version<'3.4'",
-          "csbdeep>=0.6.0,<0.7.0",
+          "csbdeep>=0.7.2",
```
Comment (tibuch) [inline setup.py:50]: Can we pin it to be <0.8.0?

**Review by tibuch — CHANGES_REQUESTED** [review]
Thank you @jdeschamps for cleaning up! 

So far I only glanced over it and it looks all reasonable to me. I will check it out and run the tests later today.

File: setup.py (hunk 1 above)
Comment (jdeschamps) [reply setup.py]: Yes that would make sense. Done in https://github.com/juglab/n2v/pull/127/commits/63412f03fa60ec68e4acb5e92e108391c3d6e787.

**Comment by jdeschamps** [comment]
We also need to update the installation instruction. I will test this later today!

[This page](https://www.tensorflow.org/install/pip) recommends to also specify `cudnn` version during installation.

**Comment by tibuch** [comment]
It is probably a good idea to follow this recommendation.

Could you ping me once you updated the instructions? Then I can test-run it from start to finish :v:

**Comment by jdeschamps** [comment]
> It is probably a good idea to follow this recommendation.
> 
> Could you ping me once you updated the instructions? Then I can test-run it from start to finish v

Done (https://github.com/juglab/n2v/pull/127/commits/0c0f7905e97481f43412364668127e52d5cbcf84).

File: README.md [hunk 2]
```diff
@@ -13,22 +13,40 @@ Our implementation is based on [CSBDEEP](http://csbdeep.bioimagecomputing.com) (
...
+
+**For linux:**
 ```
-conda create -n n2v python=3.8 cudatoolkit=11.0 cudnn=8.0 -c conda-forge
+conda install -n n2v -c conda-forge cudatoolkit=11.2 cudnn=8.1.0
 conda activate n2v
-pip install tensorflow==2.4
+
+# export path to be able to use GPU
+export LD_LIBRARY_PATH=$LD_LIBRARY_PATH:$CONDA_PREFIX/lib/
```
Comment (tibuch) [inline README.md:37]: Should we add this convenient function:
```
mkdir -p $CONDA_PREFIX/etc/conda/activate.d
echo 'export LD_LIBRARY_PATH=$LD_LIBRARY_PATH:$CONDA_PREFIX/lib/' > $CONDA_PREFIX/etc/conda/activate.d/env_vars.sh
```
Taken from the tensorflow install instructions. 
Con: It hides away that the LD_LIBRARY_PATH is set.
Pro: It automatically sets the LD_LIBRARY_PATH up on env activation.

File: README.md [hunk 3]
```diff
@@ -13,22 +13,40 @@ Our implementation is based on [CSBDEEP](http://csbdeep.bioimagecomputing.com) (
...
+CuDNN installed on your machine.
+_Note:_ Macs are not GPU-supported with TensorFlow.
 
+The [Tensorflow guidelines](https://www.tensorflow.org/install/pip) provide a step by step guide on how to install
+TensorFlow on the various operating system. 
+
+**For linux:**
 ```
-conda create -n n2v python=3.8 cudatoolkit=11.0 cudnn=8.0 -c conda-forge
+conda install -n n2v -c conda-forge cudatoolkit=11.2 cudnn=8.1.0
```
Comment (tibuch) [inline README.md:33]: Why did you change `conda create -n ...` to `conda install -n ...`? 

`conda install ...` does not work for me.

File: README.md [hunk 4]
```diff
@@ -13,22 +13,40 @@ Our implementation is based on [CSBDEEP](http://csbdeep.bioimagecomputing.com) (
...
 #### If you start from scratch...
 We recommend using [miniconda](https://docs.conda.io/en/latest/miniconda.html).
 If you do not yet have a strong opinion, just use it too!
 
-After installing Miniconda, the following lines might are likely the easiest way to get Tensorflow and CuDNN installed on your machine (_Note:_ Macs are not supported, and if you sit on a Windows machine all this might also require some modifications.):
+After installing Miniconda, the following lines might are likely the easiest way to get Tensorflow and 
+CuDNN installed on your machine.
+_Note:_ Macs are not GPU-supported with TensorFlow.
 
+The [Tensorflow guidelines](https://www.tensorflow.org/install/pip) provide a step by step guide on how to install
```
Comment (tibuch) [inline README.md:28]: Maybe we can point directly to the step-by-step instructions: https://www.tensorflow.org/install/pip#step-by-step_instructions

In that case I would also vote to remove any install instructions for tensorflow. I was kind of confused and started following the **For linux:** instructions :innocent:

**Comment by tibuch** [comment]
Thank you @jdeschamps for all the work! 

I was able to install everything and run the tests and example notebooks as they are. 

I did __not__ run the functional test. They just take too long.

I also tested the N2V2 functionality by setting 
* `unet_residual = False`
* `skip_skipone = True`
* `blurpool = True`
* `n2v_manipulator = 'median'`

Training and prediction works well for the 2D case (in 3D the model is not initialized with an `NotImplementedError` #129). However, `model.export` does not work with the current `blurpool` implementation (see #128).

**Comment by jdeschamps** [comment]
@tibuch oh good catch. 

I have also not tried the functional tests, I have excluded them from tox testenv. I will start them on a virtual machine today and check them tomorrow.

One more thing. It seems that on my RedHat machine, I need to set the `export LD_LIBRARY_PATH=$LD_LIBRARY_PATH:$CONDA_PREFIX/lib/` every time I start the environment new (e.g.: you close the terminal and restart it). Do you also have this or is it RedHat specific? It may be due to newer versions of TF...
If that is the case, that is super annoying and we should consider adding the `export` to the notebooks.

**Comment by tibuch** [comment]
This should be fixed by running this once in the acitvate env:
```
mkdir -p $CONDA_PREFIX/etc/conda/activate.d
echo 'export LD_LIBRARY_PATH=$LD_LIBRARY_PATH:$CONDA_PREFIX/lib/' > $CONDA_PREFIX/etc/conda/activate.d/env_vars.sh
```

File: README.md (hunk 4 above)
Comment (jdeschamps) [reply README.md:28]: Yes that would make more sense

**Comment by jdeschamps** [comment]
> This should be fixed by running this once in the acitvate env:
> 
> ```
> mkdir -p $CONDA_PREFIX/etc/conda/activate.d
> echo 'export LD_LIBRARY_PATH=$LD_LIBRARY_PATH:$CONDA_PREFIX/lib/' > $CONDA_PREFIX/etc/conda/activate.d/env_vars.sh
> ```

I actually had missed your comments because I didn't refresh the page. Perfect!

File: README.md (hunk 3 above)
Comment (jdeschamps) [reply README.md:33]: That's a brain fart, merged two commands without noticing!

File: README.md (hunk 2 above)
Comment (jdeschamps) [reply README.md:37]: Much much better than having the `export` everywhere...

**Comment by jdeschamps** [comment]
The functional tests run through without errors btw.

**Comment by jdeschamps** [comment]
I merged here the fix for the model export (https://github.com/juglab/n2v/issues/128). However there is still the bug that we cannot save twice in a row. I believe this is due to the way CSBDeep saves the model to `.pb` format (see https://github.com/CSBDeep/CSBDeep/issues/62).

Since the export works, I would suggest to create an issue about this bug to keep track of it and see what becomes of it in the next CSBDeep version. In the mean time since it does not prevent users from using N2V notebooks, I would go ahead with the new version release.
